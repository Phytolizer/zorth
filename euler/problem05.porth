include "std.porth"

macro N 20 end

// |acc          |tmp           |ans|
macro acc mem end
macro tmp mem 8 N * + end
macro ans tmp 8 N * + end

2 while dup N <= do

  // clean up the tmp table
  0 while dup N < do
    tmp over 8 * + 0 .64
    1 +
  end drop

  // factorize
  dup 2 while over 1 > do
    2dup mod 0 = if
      dup 8 * tmp + inc64
      swap over / swap
    else
      1 +
    end
  end 2drop

  0 while dup N < do
    dup 8 * acc + ,64
    over 8 * tmp + ,64
    < if
      dup 8 * acc +
      over 8 * tmp + ,64
      .64
    end

    1 +
  end drop

  1 +
end drop

ans 1 .64

0 while dup N < do
  acc over 8 * + ,64 while dup 0 > do
    over ans ,64 * ans swap .64
    1 -
  end drop

  1 +
end drop

ans ,64 print
