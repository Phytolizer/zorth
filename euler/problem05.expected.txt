: returncode 0
: stdout 10
232792560

: stderr 0

